---
title: "Supplementary Text S5: Runtime benchmark"
author:
- name: Fabricio Almeida-Silva
  affiliation: |
    VIB-UGent Center for Plant Systems Biology, Ghent University, 
    Ghent, Belgium
- name: Yves Van de Peer
  affiliation: |
    VIB-UGent Center for Plant Systems Biology, Ghent University, 
    Ghent, Belgium
output:
  BiocStyle::pdf_document
vignette: |
  %\VignetteIndexEntry{Supplementary Text S5}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    warning = FALSE, 
    eval = FALSE
)

Sys.setlocale("LC_TIME", "en_US.UTF-8")
Sys.setlocale("LC_PAPER", "en_US.UTF-8")
Sys.setlocale("LC_MEASUREMENT", "en_US.UTF-8")
```

# Introduction

Here, we will perform a runtime benchmark for functions related to duplicate 
classification and substitution rates calculation using model organisms.

To start, let's load the required data and packages.

```{r load_packages, eval = TRUE}
set.seed(123) # for reproducibility

# Load required packages
library(doubletrouble)
library(here)
library(tidyverse)
library(patchwork)

source(here("code", "utils.R"))

# Load sample metadata for Ensembl instances
load(here("products", "result_files", "metadata_all.rda"))
```

# Benchmark 1: `classify_gene_pairs()`

Here, we will benchmark the performance of `classify_gene_pairs()`
with model organisms.

First, let's get the genome and annotation data.

```{r}
# Create a data frame with names of model species and their Ensembl instances
model_species <- data.frame(
    species = c(
        "arabidopsis_thaliana", "caenorhabditis_elegans", 
        "homo_sapiens", "saccharomyces_cerevisiae",
        "drosophila_melanogaster", "danio_rerio"
    ),
    instance = c(
        "plants", "metazoa", "ensembl", "fungi", "metazoa", "ensembl"
    )
)

# For each organism, download data, and identify and classify duplicates
model_duplicates <- lapply(seq_len(nrow(model_species)), function(x) {
    
    species <- model_species$species[x]
    instance <- model_species$instance[x]
    
    # Get annotation
    annot <- get_annotation(model_species[x, ], instance)
    
    # Get proteome and keep only primary transcripts
    seq <- get_proteomes(model_species[x, ], instance)
    seq <- filter_sequences(seq, annot)
    
    # Process data
    pdata <- syntenet::process_input(seq, annot, filter_annotation = TRUE)
    
    # Perform DIAMOND search
    outdir <- file.path(tempdir(), paste0(species, "_intra"))
    diamond <- syntenet::run_diamond(
        seq = pdata$seq,
        compare = "intraspecies", 
        outdir = outdir,
        threads = 4,
        ... = "--sensitive"
    )
    
    fs::dir_delete(outdir)
    
    # Classify duplicates - standard mode
    start <- Sys.time()
    duplicate_pairs <- classify_gene_pairs(
            blast_list = diamond,
            annotation = pdata$annotation,
            scheme = "standard"
        )[[1]]
    end <- Sys.time()
    runtime <- end - start
    
    return(runtime)
})
names(model_duplicates) <- gsub("_", " ", str_to_title(model_species$species))

# Summarize results in a table
benchmark_classification <- data.frame(
    species = names(model_duplicates),
    time_seconds = as.numeric(unlist(model_duplicates))
)

# Save results
save(
    benchmark_classification, compress = "xz",
    file = here("products", "result_files", "benchmark_classification.rda")
)
```

# Benchmark 2: `pairs2kaks()`

Next, we will benchmark the performance of `pairs2kaks()` for
duplicate pairs in the *Saccharomyces cerevisiae* genome.
We will do it using a single thread, and using parallelization (with
4 and 8 threads).

First of all, let's get the required data for `pairs2kaks()`.

```{r}
# Load duplicate pairs for S. cerevisiae
load(here("products", "result_files", "fungi_duplicates.rda"))
scerevisiae_pairs <- fungi_duplicates["saccharomyces_cerevisiae"]

# Get CDS for S. cerevisiae
scerevisiae_cds <- get_cds_ensembl("saccharomyces_cerevisiae", "fungi")
```

Now, we can do the benchmark. 

```{r}
# Parallel back-end: SerialParam (1 thread)
start <- Sys.time()
kaks <- pairs2kaks(
    scerevisiae_pairs, 
    scerevisiae_cds,
    bp_param = BiocParallel::SerialParam()
)
end <- Sys.time()
runtime_serial <- end - start

# Parallel back-end: SnowParam, 4 threads
start <- Sys.time()
kaks <- pairs2kaks(
    scerevisiae_pairs, 
    scerevisiae_cds,
    bp_param = BiocParallel::SnowParam(workers = 4)
)
end <- Sys.time()
runtime_snow4 <- end - start

# Parallel back-end: SnowParam, 8 threads
start <- Sys.time()
kaks <- pairs2kaks(
    scerevisiae_pairs, 
    scerevisiae_cds,
    bp_param = BiocParallel::SnowParam(workers = 8)
)
end <- Sys.time()
runtime_snow8 <- end - start

# Summarize results in a table
benchmark_kaks <- data.frame(
    `Back-end` = c("Serial", "Snow, 4 threads", "Snow, 8 threads"),
    Time_minutes = as.numeric(c(runtime_serial, runtime_snow4, runtime_snow8))
) |>
    dplyr::mutate(
        Pairs_per_minute = nrow(scerevisiae_pairs[[1]]) / Time_minutes,
        Pairs_per_second = nrow(scerevisiae_pairs[[1]]) / (Time_minutes * 60)
    )

save(
    benchmark_kaks, compress = "xz",
    file = here("products", "result_files", "benchmark_kaks.rda")
)
```

# Session info {.unnumbered}

This document was created under the following conditions:

```{r sessionInfo, echo = FALSE, eval = TRUE}
sessioninfo::session_info()
```

