<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Code and data for “doubletrouble: an R/Bioconductor package for the identification, classification, and analyses of gene and genome duplications” - 1&nbsp; Obtaining species trees for Ensembl instances</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/chapter_02.html" rel="next">
<link href="../index.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/chapter_01.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Obtaining species trees for Ensembl instances</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Code and data for “doubletrouble: an R/Bioconductor package for the identification, classification, and analyses of gene and genome duplications”</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/chapter_01.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Obtaining species trees for Ensembl instances</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/chapter_02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Identification and classification of duplicated genes in Ensembl and Ensembl Genomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/chapter_03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Calculating substitution rates for selected Ensembl genomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/chapter_04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visual exploration of duplicated genes across the Eukarya tree of life</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/chapter_05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Runtime benchmark</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#summary-stats" id="toc-summary-stats" class="nav-link active" data-scroll-target="#summary-stats"><span class="header-section-number">1.1</span> Summary stats</a></li>
  <li><a href="#getting-species-metadata" id="toc-getting-species-metadata" class="nav-link" data-scroll-target="#getting-species-metadata"><span class="header-section-number">1.2</span> Getting species metadata</a></li>
  <li><a href="#busco-guided-phylogeny-inference" id="toc-busco-guided-phylogeny-inference" class="nav-link" data-scroll-target="#busco-guided-phylogeny-inference"><span class="header-section-number">1.3</span> BUSCO-guided phylogeny inference</a>
  <ul class="collapse">
  <li><a href="#obtaining-busco-sequences" id="toc-obtaining-busco-sequences" class="nav-link" data-scroll-target="#obtaining-busco-sequences"><span class="header-section-number">1.3.1</span> Obtaining BUSCO sequences</a></li>
  <li><a href="#tree-inference-from-busco-genes" id="toc-tree-inference-from-busco-genes" class="nav-link" data-scroll-target="#tree-inference-from-busco-genes"><span class="header-section-number">1.3.2</span> Tree inference from BUSCO genes</a></li>
  </ul></li>
  <li><a href="#obtaining-busco-scores" id="toc-obtaining-busco-scores" class="nav-link" data-scroll-target="#obtaining-busco-scores"><span class="header-section-number">1.4</span> Obtaining BUSCO scores</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Obtaining species trees for Ensembl instances</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Here, we will describe the code to obtain a species tree for each Ensembl instance using BUSCO genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Herper)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(taxize)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cogeqc)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># for reproducibility</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="fl">1e6</span>) <span class="co"># to allow download of big files</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">here</span>(<span class="st">"code"</span>, <span class="st">"utils.R"</span>))</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">here</span>(<span class="st">"code"</span>, <span class="st">"utils_busco_phylogeny.R"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="summary-stats" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="summary-stats"><span class="header-section-number">1.1</span> Summary stats</h2>
<p>To start with, let’s get the number of species for each instance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get number of species in Ensembl Genomes</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>instances <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fungi_mart"</span>, <span class="st">"plants_mart"</span>, <span class="st">"metazoa_mart"</span>, <span class="st">"protists_mart"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>nspecies_ensemblgenomes <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(instances, <span class="cf">function</span>(x) {</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">nrow</span>(<span class="fu">listDatasets</span>(<span class="fu">useEnsemblGenomes</span>(<span class="at">biomart =</span> x))))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get number of species in Ensembl</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>nspecies_ensembl <span class="ot">&lt;-</span> <span class="fu">nrow</span>(<span class="fu">listDatasets</span>(<span class="fu">useEnsembl</span>(<span class="at">biomart =</span> <span class="st">"genes"</span>)))</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine summary stats onto a data frame</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>nspecies_all <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">instance =</span> <span class="fu">c</span>(<span class="fu">gsub</span>(<span class="st">"_mart"</span>, <span class="st">""</span>, instances), <span class="st">"ensembl"</span>),</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_genes =</span> <span class="fu">c</span>(nspecies_ensemblgenomes, nspecies_ensembl)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>nspecies_all</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  instance n_genes
1    fungi      70
2   plants     151
3  metazoa     280
4 protists      33
5  ensembl     214</code></pre>
</div>
</div>
</section>
<section id="getting-species-metadata" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="getting-species-metadata"><span class="header-section-number">1.2</span> Getting species metadata</h2>
<p>Now, let’s get species metadata for each Ensembl instance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Metadata column names</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>col_names <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"name"</span>, <span class="st">"species"</span>, <span class="st">"division"</span>, <span class="st">"taxonomy_id"</span>, <span class="st">"assembly"</span>, </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"assembly_accession"</span>, <span class="st">"genebuild"</span>, <span class="st">"variation"</span>, <span class="st">"microarray"</span>, <span class="st">"pan_compara"</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"peptide_compara"</span>, <span class="st">"genome_alignments"</span>, <span class="st">"other_alignments"</span>, <span class="st">"core_db"</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"species_id"</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>to_remove <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"variation"</span>, <span class="st">"microarray"</span>, <span class="st">"pan_compara"</span>, <span class="st">"peptide_compara"</span>, </span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"genome_alignments"</span>, <span class="st">"other_alignments"</span>, <span class="st">"core_db"</span>, <span class="st">"species_id"</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensembl Fungi</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>metadata_fungi <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-57/fungi/species_EnsemblFungi.txt"</span>,</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> col_names, <span class="at">skip =</span> <span class="dv">1</span>, <span class="at">col_select =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">startsWith</span>(core_db, <span class="st">"fungi_"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">any_of</span>(to_remove)) <span class="sc">|&gt;</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>metadata_fungi <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>    metadata_fungi,</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">classification</span>(metadata_fungi<span class="sc">$</span>taxonomy_id, <span class="at">db =</span> <span class="st">"ncbi"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>        <span class="fu">format_classification</span>()</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensembl Plants</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>metadata_plants <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>    <span class="st">"http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-57/plants/species_EnsemblPlants.txt"</span>,</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> col_names, <span class="at">skip =</span> <span class="dv">1</span>, <span class="at">col_select =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(species <span class="sc">!=</span> <span class="st">"triticum_aestivum_kariega"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">any_of</span>(to_remove)) <span class="sc">|&gt;</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>metadata_plants <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>    metadata_plants,</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">classification</span>(metadata_plants<span class="sc">$</span>taxonomy_id, <span class="at">db =</span> <span class="st">"ncbi"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">format_classification</span>()</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensembl Metazoa</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>metadata_metazoa <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>    <span class="st">"http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-57/metazoa/species_EnsemblMetazoa.txt"</span>,</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> col_names, <span class="at">skip =</span> <span class="dv">1</span>, <span class="at">col_select =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">startsWith</span>(core_db, <span class="st">"metazoa_"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">any_of</span>(to_remove)) <span class="sc">|&gt;</span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>metadata_metazoa <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a>    metadata_metazoa,</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">classification</span>(metadata_metazoa<span class="sc">$</span>taxonomy_id, <span class="at">db =</span> <span class="st">"ncbi"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a>        <span class="fu">format_classification</span>()</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensembl Protists</span></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>metadata_protists <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>    <span class="st">"http://ftp.ebi.ac.uk/ensemblgenomes/pub/release-57/protists/species_EnsemblProtists.txt"</span>,</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> col_names, <span class="at">skip =</span> <span class="dv">1</span>, <span class="at">col_select =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">startsWith</span>(core_db, <span class="st">"protists_"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">any_of</span>(to_remove)) <span class="sc">|&gt;</span></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>metadata_protists <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>    metadata_protists,</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>    <span class="fu">classification</span>(metadata_protists<span class="sc">$</span>taxonomy_id, <span class="at">db =</span> <span class="st">"ncbi"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>        <span class="fu">format_classification</span>()</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensembl</span></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>metadata_ensembl <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(</span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://ftp.ensembl.org/pub/release-110/species_EnsemblVertebrates.txt"</span>,</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> col_names, <span class="at">skip =</span> <span class="dv">1</span>, <span class="at">col_select =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">any_of</span>(to_remove)) <span class="sc">|&gt;</span></span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>metadata_ensembl <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a>    metadata_ensembl,</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">classification</span>(metadata_ensembl<span class="sc">$</span>taxonomy_id, <span class="at">db =</span> <span class="st">"ncbi"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>        <span class="fu">format_classification</span>()</span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a><span class="co"># Combining all metadata data frames into a list and saving it</span></span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>metadata_all <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">fungi =</span> metadata_fungi,</span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a>    <span class="at">plants =</span> metadata_plants,</span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a>    <span class="at">metazoa =</span> metadata_metazoa,</span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a>    <span class="at">protists =</span> metadata_protists,</span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a>    <span class="at">ensembl =</span> metadata_ensembl</span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-97"><a href="#cb4-97" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb4-98"><a href="#cb4-98" aria-hidden="true" tabindex="-1"></a>    metadata_all, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb4-99"><a href="#cb4-99" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"metadata_all.rda"</span>)</span>
<span id="cb4-100"><a href="#cb4-100" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="busco-guided-phylogeny-inference" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="busco-guided-phylogeny-inference"><span class="header-section-number">1.3</span> BUSCO-guided phylogeny inference</h2>
<p>Here, for each Ensembl instance, we infer a species tree using the following workflow:</p>
<ol type="1">
<li>Run BUSCO in protein mode with <strong>cogeqc</strong>, using translated sequences for primary transcripts as input;</li>
<li>Get the sequences of the identified complete BUSCOs that are shared across all species;</li>
<li>Perform a multiple sequence alignment for each BUSCO gene family.</li>
<li>Trim the alignments to remove columns with &gt;50% of gaps.</li>
<li>Infer a phylogeny with IQ-TREE2.</li>
</ol>
<p>To start with, we will use the Bioconductor package <strong>Herper</strong> to create a Conda environment containing BUSCO and all its dependencies. Then, we will use this environment to run BUSCO from the R session.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Conda environment with BUSCO</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>my_miniconda <span class="ot">&lt;-</span> <span class="st">"~/"</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>conda <span class="ot">&lt;-</span> <span class="fu">install_CondaTools</span>(</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tools =</span> <span class="st">"busco==5.5.0"</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">env =</span> <span class="st">"busco_env"</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathToMiniConda =</span> my_miniconda</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="obtaining-busco-sequences" class="level3" data-number="1.3.1">
<h3 data-number="1.3.1" class="anchored" data-anchor-id="obtaining-busco-sequences"><span class="header-section-number">1.3.1</span> Obtaining BUSCO sequences</h3>
<p>To obtain sequences for BUSCO genes, we will run BUSCO in protein mode using the R/Bioconductor package <strong>cogeqc</strong>. Then, we will read the sequences for complete, single-copy BUSCOs, and keep only BUSCO genes that are shared by a certain % of the species. Ideally, this cut-off should be 100% of conservation (i.e., the BUSCO gene is found in all species), but it can be relaxed for some clades.</p>
<section id="ensembl-fungi" class="level4" data-number="1.3.1.1">
<h4 data-number="1.3.1.1" class="anchored" data-anchor-id="ensembl-fungi"><span class="header-section-number">1.3.1.1</span> Ensembl Fungi</h4>
<p>Here, we will obtain BUSCO genes for Ensembl Fungi species using the following parameters:</p>
<ol type="1">
<li>Lineage: <strong>eukaryota_odb10</strong></li>
<li>Conservation: 100%</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download whole-genome protein sequences to a directory sequences</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>busco_fungi <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"~/Downloads/busco_fungi"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>seq_fungi <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_fungi, <span class="st">"seqs"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(seq_fungi)) { <span class="fu">dir.create</span>(seq_fungi, <span class="at">recursive =</span> <span class="cn">TRUE</span>) }</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download_filtered_proteomes</span>(metadata_all<span class="sc">$</span>fungi, <span class="st">"fungi"</span>, seq_fungi)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run BUSCO in `protein` mode</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="fu">with_CondaEnv</span>(</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"busco_env"</span>,</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    cogeqc<span class="sc">::</span><span class="fu">run_busco</span>(</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequence =</span> seq_fungi, </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">outlabel =</span> <span class="st">"ensembl_fungi"</span>,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">mode =</span> <span class="st">"protein"</span>,</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">lineage =</span> <span class="st">"eukaryota_odb10"</span>,</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">outpath =</span> busco_fungi,</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">threads =</span> <span class="dv">3</span>,</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">download_path =</span> busco_fungi</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathToMiniConda =</span> my_miniconda</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_fungi, <span class="st">"ensembl_fungi"</span>)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>fungi_busco_seqs <span class="ot">&lt;-</span> <span class="fu">read_busco_sequences</span>(outdir, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Saving BUSCO sequences:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save list of AAStringSet objects with conserved BUSCO sequences</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    fungi_busco_seqs, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_seqs"</span>, <span class="st">"fungi_busco_seqs.rda"</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-plants" class="level4" data-number="1.3.1.2">
<h4 data-number="1.3.1.2" class="anchored" data-anchor-id="ensembl-plants"><span class="header-section-number">1.3.1.2</span> Ensembl Plants</h4>
<p>Here, we will use the lineage data set <strong>eukaryota_odb10</strong>. We could use <strong>viridiplantae_odb10</strong>, but there are 3 Rhodophyta species (<em>Chondrus crispus</em>, <em>Galdieria sulphuraria</em>, and <em>Cyanidioschyzon merolae</em>). Because none of the BUSCO genes were shared by all species, we selected genes shared by &gt;60% of the species, and then manually selected BUSCO genes in a way that all species are included. This was required because some taxa (in particular <em>Triticum</em> species) had very few BUSCO genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download whole-genome protein sequences to a directory sequences</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>busco_plants <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"~/Downloads/busco_plants"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>seq_plants <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_plants, <span class="st">"seqs"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(seq_plants)) { <span class="fu">dir.create</span>(seq_plants, <span class="at">recursive =</span> <span class="cn">TRUE</span>) }</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download_filtered_proteomes</span>(metadata_all<span class="sc">$</span>plants, <span class="st">"plants"</span>, seq_plants)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run BUSCO in `protein` mode</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="fu">with_CondaEnv</span>(</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"busco_env"</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    cogeqc<span class="sc">::</span><span class="fu">run_busco</span>(</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequence =</span> seq_plants, </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">outlabel =</span> <span class="st">"ensemblplants"</span>,</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">mode =</span> <span class="st">"protein"</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">lineage =</span> <span class="st">"eukaryota_odb10"</span>,</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">outpath =</span> busco_plants,</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">threads =</span> <span class="dv">4</span>,</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">download_path =</span> busco_plants</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathToMiniConda =</span> my_miniconda</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Read sequences of BUSCOs preserved in &gt;=60% of the species</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_plants, <span class="st">"ensemblplants"</span>)</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>plants_busco_seqs <span class="ot">&lt;-</span> <span class="fu">read_busco_sequences</span>(outdir, <span class="at">conservation_freq =</span> <span class="fl">0.6</span>)</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 10 BUSCO genes so that all species are represented</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>plants_busco_pav <span class="ot">&lt;-</span> <span class="fu">get_busco_pav</span>(plants_busco_seqs)</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a><span class="co">#' The following code was used to manually select BUSCOs in a way that</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a><span class="co">#' all species are represented</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ht &lt;- ComplexHeatmap::Heatmap(plants_busco_pav)</span></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ht &lt;- ComplexHeatmap::draw(ht)</span></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; InteractiveComplexHeatmap::htShiny(ht)</span></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector of selected BUSCOs</span></span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>selected_buscos <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>    <span class="st">"549762at2759"</span>, <span class="st">"1003258at2759"</span>, <span class="st">"1247641at2759"</span>,</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1200489at2759"</span>, <span class="st">"1398309at2759"</span>, <span class="st">"1346432at2759"</span>,</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1266231at2759"</span>, <span class="st">"1094121at2759"</span>, <span class="st">"1421503at2759"</span>,</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>    <span class="st">"664730at2759"</span>, <span class="st">"1405073at2759"</span>, <span class="st">"450058at2759"</span>,</span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>    <span class="st">"865202at2759"</span>, <span class="st">"901894at2759"</span>, <span class="st">"1450538at2759"</span>,</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1284731at2759"</span></span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset sequences to keep only selected BUSCOs</span></span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>plants_busco_seqs <span class="ot">&lt;-</span> plants_busco_seqs[selected_buscos]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Saving BUSCO sequences:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save list of AAStringSet objects with conserved BUSCO sequences</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    plants_busco_seqs, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_seqs"</span>, <span class="st">"plants_busco_seqs.rda"</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-protists" class="level4" data-number="1.3.1.3">
<h4 data-number="1.3.1.3" class="anchored" data-anchor-id="ensembl-protists"><span class="header-section-number">1.3.1.3</span> Ensembl Protists</h4>
<p>Here, we will obtain BUSCO genes for Ensembl Protists species using the following parameters:</p>
<ol type="1">
<li>Lineage: <strong>eukaryota_odb10</strong></li>
<li>Conservation: 100%</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download whole-genome protein sequences to a directory sequences</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>busco_protists <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"~/Downloads/busco_protists"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>seq_protists <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_protists, <span class="st">"seqs"</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(seq_protists)) { <span class="fu">dir.create</span>(seq_protists, <span class="at">recursive =</span> <span class="cn">TRUE</span>) }</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download_filtered_proteomes</span>(metadata_all<span class="sc">$</span>protists, <span class="st">"protists"</span>, seq_protists)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run BUSCO in `protein` mode</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="fu">with_CondaEnv</span>(</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"busco_env"</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    cogeqc<span class="sc">::</span><span class="fu">run_busco</span>(</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequence =</span> seq_protists, </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">outlabel =</span> <span class="st">"ensemblprotists"</span>,</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">mode =</span> <span class="st">"protein"</span>,</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">lineage =</span> <span class="st">"eukaryota_odb10"</span>,</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">outpath =</span> busco_protists,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">threads =</span> <span class="dv">4</span>,</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">download_path =</span> busco_protists</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathToMiniConda =</span> my_miniconda</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Read sequences of BUSCOs preserved in &gt;=60% of the species</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_protists, <span class="st">"ensemblprotists"</span>)</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>protists_busco_seqs <span class="ot">&lt;-</span> <span class="fu">read_busco_sequences</span>(outdir, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Saving BUSCO sequences:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save list of AAStringSet objects with conserved BUSCO sequences</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    protists_busco_seqs, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_seqs"</span>, <span class="st">"protists_busco_seqs.rda"</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-metazoa" class="level4" data-number="1.3.1.4">
<h4 data-number="1.3.1.4" class="anchored" data-anchor-id="ensembl-metazoa"><span class="header-section-number">1.3.1.4</span> Ensembl Metazoa</h4>
<p>For the Metazoa instance, we used the <em>metazoa_odb10</em> lineage data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download whole-genome protein sequences to a directory sequences</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>busco_metazoa <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"~/Downloads/busco_metazoa"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>seq_metazoa <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_metazoa, <span class="st">"seqs"</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(seq_metazoa)) { <span class="fu">dir.create</span>(seq_metazoa, <span class="at">recursive =</span> <span class="cn">TRUE</span>) }</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download_filtered_proteomes</span>(metadata_all<span class="sc">$</span>metazoa, <span class="st">"metazoa"</span>, seq_metazoa)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run BUSCO in `protein` mode</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">with_CondaEnv</span>(</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"busco_env"</span>,</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    cogeqc<span class="sc">::</span><span class="fu">run_busco</span>(</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequence =</span> seq_metazoa, </span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">outlabel =</span> <span class="st">"ensemblmetazoa"</span>,</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">mode =</span> <span class="st">"protein"</span>,</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">lineage =</span> <span class="st">"metazoa_odb10"</span>,</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">outpath =</span> busco_metazoa,</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">threads =</span> <span class="dv">4</span>,</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">download_path =</span> busco_metazoa</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathToMiniConda =</span> my_miniconda</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Read sequences of BUSCOs preserved in &gt;=60% of the species</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_metazoa, <span class="st">"ensemblmetazoa"</span>)</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>metazoa_busco_seqs <span class="ot">&lt;-</span> <span class="fu">read_busco_sequences</span>(outdir, <span class="at">conservation_freq =</span> <span class="fl">0.9</span>)</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 10 BUSCO genes so that all species are represented</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>metazoa_busco_pav <span class="ot">&lt;-</span> <span class="fu">get_busco_pav</span>(metazoa_busco_seqs)</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a><span class="co">#' The following code was used to manually select BUSCOs in a way that</span></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a><span class="co">#' all species are represented</span></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ht &lt;- ComplexHeatmap::Heatmap(metazoa_busco_pav)</span></span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ht &lt;- ComplexHeatmap::draw(ht)</span></span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; InteractiveComplexHeatmap::htShiny(ht)</span></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector of selected BUSCOs</span></span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>selected_buscos <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>    <span class="st">"351226at33208"</span>, <span class="st">"135294at33208"</span>,</span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>    <span class="st">"517525at33208"</span>, <span class="st">"501396at33208"</span>,</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>    <span class="st">"464987at33208"</span>, <span class="st">"443518at33208"</span>,</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>    <span class="st">"495100at33208"</span>, <span class="st">"335107at33208"</span>,</span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>    <span class="st">"454911at33208"</span>, <span class="st">"134492at33208"</span></span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset sequences to keep only selected BUSCOs</span></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a>metazoa_busco_seqs <span class="ot">&lt;-</span> metazoa_busco_seqs[selected_buscos]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Saving BUSCO sequences:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save list of AAStringSet objects with conserved BUSCO sequences</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    metazoa_busco_seqs, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_seqs"</span>, <span class="st">"metazoa_busco_seqs.rda"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-vertebrates" class="level4" data-number="1.3.1.5">
<h4 data-number="1.3.1.5" class="anchored" data-anchor-id="ensembl-vertebrates"><span class="header-section-number">1.3.1.5</span> Ensembl Vertebrates</h4>
<p>Here, because there are 3 non-vertebrate species (<em>C. elegans</em>, <em>D. melanogaster</em>, and <em>S. cerevisiae</em>), we will use the lineage data set <strong>eukaryota_odb10</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download whole-genome protein sequences to a directory sequences</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>busco_vertebrates <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"~/Downloads/busco_vertebrates"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>seq_vertebrates <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_vertebrates, <span class="st">"seqs"</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(seq_vertebrates)) { <span class="fu">dir.create</span>(seq_vertebrates, <span class="at">recursive =</span> <span class="cn">TRUE</span>) }</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download_filtered_proteomes</span>(metadata_all<span class="sc">$</span>ensembl, <span class="st">"ensembl"</span>, seq_vertebrates)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run BUSCO in `protein` mode</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">with_CondaEnv</span>(</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"busco_env"</span>,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    cogeqc<span class="sc">::</span><span class="fu">run_busco</span>(</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">sequence =</span> seq_vertebrates, </span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">outlabel =</span> <span class="st">"ensemblvertebrates"</span>,</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">mode =</span> <span class="st">"protein"</span>,</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">lineage =</span> <span class="st">"eukaryota_odb10"</span>,</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">outpath =</span> busco_vertebrates,</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">threads =</span> <span class="dv">4</span>,</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">download_path =</span> busco_vertebrates</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathToMiniConda =</span> my_miniconda</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Read sequences of BUSCOs preserved in &gt;=90% of the species</span></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_vertebrates, <span class="st">"ensemblvertebrates"</span>)</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>vertebrates_busco_seqs <span class="ot">&lt;-</span> <span class="fu">read_busco_sequences</span>(outdir, <span class="at">conservation_freq =</span> <span class="fl">0.9</span>)</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 10 BUSCO genes so that all species are represented</span></span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>vertebrates_busco_pav <span class="ot">&lt;-</span> <span class="fu">get_busco_pav</span>(vertebrates_busco_seqs)</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a><span class="co">#' The following code was used to manually select BUSCOs in a way that</span></span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a><span class="co">#' all species are represented</span></span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ht &lt;- ComplexHeatmap::Heatmap(vertebrates_busco_pav)</span></span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ht &lt;- ComplexHeatmap::draw(ht)</span></span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; InteractiveComplexHeatmap::htShiny(ht)</span></span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector of selected BUSCOs</span></span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>selected_buscos <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a>    <span class="st">"834694at2759"</span>, <span class="st">"551907at2759"</span>,</span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a>    <span class="st">"491869at2759"</span>, <span class="st">"1085752at2759"</span>,</span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a>    <span class="st">"801857at2759"</span>, <span class="st">"1398309at2759"</span>,</span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a>    <span class="st">"176625at2759"</span>, <span class="st">"1324510at2759"</span>,</span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1377237at2759"</span>, <span class="st">"1085752at2759"</span></span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset sequences to keep only selected BUSCOs</span></span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true" tabindex="-1"></a>vertebrates_busco_seqs <span class="ot">&lt;-</span> vertebrates_busco_seqs[selected_buscos]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Saving BUSCO sequences:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save list of AAStringSet objects with conserved BUSCO sequences</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    vertebrates_busco_seqs, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_seqs"</span>, <span class="st">"vertebrates_busco_seqs"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="tree-inference-from-busco-genes" class="level3" data-number="1.3.2">
<h3 data-number="1.3.2" class="anchored" data-anchor-id="tree-inference-from-busco-genes"><span class="header-section-number">1.3.2</span> Tree inference from BUSCO genes</h3>
<p>Now, we will infer species trees from MSAs for each family, and from a single concatenated MSA (when possible).</p>
<section id="ensembl-fungi-1" class="level4" data-number="1.3.2.1">
<h4 data-number="1.3.2.1" class="anchored" data-anchor-id="ensembl-fungi-1"><span class="header-section-number">1.3.2.1</span> Ensembl Fungi</h4>
<p>Performing MSA with MAFFT and trimming the alignment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform MSA with MAFFT</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>aln_fungi <span class="ot">&lt;-</span> <span class="fu">align_sequences</span>(busco_seqs_fungi, <span class="at">threads =</span> <span class="dv">4</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim alignment to remove columns with &gt;50% of gaps</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>aln_fungi_trimmed <span class="ot">&lt;-</span> <span class="fu">lapply</span>(aln_fungi, trim_alignment, <span class="at">max_gap =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s infer a species tree using IQ-TREE2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"aphanomyces.astaci,aphanomyces.invadans,globisporangium.ultimum"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>trees_fungi <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(aln_fungi_trimmed, outgroup, <span class="at">threads =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, for comparative reasons, we will also infer a single tree from a concatenated multiple sequence alignment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Concatenate alignments</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>aln_fungi_conc <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(xscat, aln_fungi_trimmed)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(aln_fungi_conc) <span class="ot">&lt;-</span> <span class="fu">names</span>(aln_fungi_trimmed[[<span class="dv">1</span>]])</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Infer tree from concatenated alignment</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>tree_fungi_conc <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">conc =</span> aln_fungi_conc), </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>    outgroup, <span class="at">threads =</span> <span class="dv">4</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Combining the trees and saving them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine trees</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>fungi_busco_trees <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    tree_fungi_conc, trees_fungi</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    fungi_busco_trees, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"trees"</span>, <span class="st">"fungi_busco_trees.rda"</span>)</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-plants-1" class="level4" data-number="1.3.2.2">
<h4 data-number="1.3.2.2" class="anchored" data-anchor-id="ensembl-plants-1"><span class="header-section-number">1.3.2.2</span> Ensembl Plants</h4>
<p>Here, because no BUSCO gene is present in all species, we will only infer a single tree from concatenated alignments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform MSA with MAFFT</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>aln_plants <span class="ot">&lt;-</span> <span class="fu">align_sequences</span>(plants_busco_seqs, <span class="at">threads =</span> <span class="dv">4</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim alignment to remove columns with &gt;50% of gaps</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>aln_plants_trimmed <span class="ot">&lt;-</span> <span class="fu">lapply</span>(aln_plants, trim_alignment, <span class="at">max_gap =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, let’s infer a species tree from a concatenated alignment. As outgroups, we’re going to use <em>Chondrus crispus</em>, <em>Galdieria sulphuraria</em>, and <em>Cyanidioschyzon merolae</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"chondrus.crispus,galdieria.sulphuraria,cyanidioschyzon.merolae"</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Concatenate alignments</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>aln_plants_conc <span class="ot">&lt;-</span> <span class="fu">concatenate_alignments</span>(aln_plants_trimmed)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Infer tree from concatenated alignment</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>plants_busco_trees <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">conc =</span> aln_plants_conc), </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    outgroup, <span class="at">threads =</span> <span class="dv">4</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Save tree</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>    plants_busco_trees, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"trees"</span>, <span class="st">"plants_busco_trees.rda"</span>)</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-protists-1" class="level4" data-number="1.3.2.3">
<h4 data-number="1.3.2.3" class="anchored" data-anchor-id="ensembl-protists-1"><span class="header-section-number">1.3.2.3</span> Ensembl Protists</h4>
<p>For this instance, two BUSCO genes were conserved across all species, so we will infer trees for each family + a tree from a concatenated alignment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform MSA with MAFFT</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>aln_protists <span class="ot">&lt;-</span> <span class="fu">align_sequences</span>(protists_busco_seqs, <span class="at">threads =</span> <span class="dv">4</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim alignment to remove columns with &gt;50% of gaps</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>aln_protists_trimmed <span class="ot">&lt;-</span> <span class="fu">lapply</span>(aln_protists, trim_alignment, <span class="at">max_gap =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s infer species trees. As outgroup, we will use Fornicata (<em>Giardia lamblia</em>) based on <a href="10.1126/sciadv.ade4973">this paper</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"giardia.lamblia"</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Path 1: a tree per BUSCO gene</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>protists_trees1 <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    aln_protists_trimmed, outgroup, <span class="at">threads =</span> <span class="dv">4</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Path 2: a single tree from a concatenated alignment</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>protists_trees2 <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">conc =</span> <span class="fu">concatenate_alignments</span>(aln_protists_trimmed)),</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    outgroup, <span class="at">threads =</span> <span class="dv">6</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine trees and save them</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>protists_busco_trees <span class="ot">&lt;-</span> <span class="fu">c</span>(protists_trees1, protists_trees2)</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>    protists_busco_trees, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"trees"</span>, <span class="st">"protists_busco_trees.rda"</span>)</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, even though we specified <em>Giardia lamblia</em>, IQ-TREE2 placed it as an ingroup. This suggests that, based on our data (BUSCO sequences), <em>Giardia lamblia</em> may not be a good outgroup.</p>
<p>Since protists are not actually a real phylogenetic group (not monophyletic), instead of digging deeper into the real phylogeny of the group and searching for a proper outgroup, we will simply use this phylogeny but acknowledging that it may not be completely accurate.</p>
</section>
<section id="ensembl-metazoa-1" class="level4" data-number="1.3.2.4">
<h4 data-number="1.3.2.4" class="anchored" data-anchor-id="ensembl-metazoa-1"><span class="header-section-number">1.3.2.4</span> Ensembl Metazoa</h4>
<p>For this instance, two BUSCO genes were conserved across all species, so we will infer trees for each family + a tree from a concatenated alignment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform MSA with MAFFT</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>aln_metazoa <span class="ot">&lt;-</span> <span class="fu">align_sequences</span>(metazoa_busco_seqs, <span class="at">threads =</span> <span class="dv">4</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim alignment to remove columns with &gt;50% of gaps</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>aln_metazoa_trimmed <span class="ot">&lt;-</span> <span class="fu">lapply</span>(aln_metazoa, trim_alignment, <span class="at">max_gap =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s infer a species tree. As outgroup, we will use the ctenophore <em>Mnemiopsis leidyi</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"mnemiopsis.leidyi"</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get a single tree from a concatenated alignment</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>metazoa_busco_trees <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">conc =</span> <span class="fu">concatenate_alignments</span>(aln_metazoa_trimmed)),</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    outgroup, <span class="at">threads =</span> <span class="dv">6</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Save tree</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    metazoa_busco_trees, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"trees"</span>, <span class="st">"metazoa_busco_trees.rda"</span>)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-vertebrates-1" class="level4" data-number="1.3.2.5">
<h4 data-number="1.3.2.5" class="anchored" data-anchor-id="ensembl-vertebrates-1"><span class="header-section-number">1.3.2.5</span> Ensembl Vertebrates</h4>
<p>For this instance, no BUSCO gene was conserved in all species. Thus, we will infer a single tree from a concatenated alignment of ten representative BUSCOs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform MSA with MAFFT</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>aln_vertebrates <span class="ot">&lt;-</span> <span class="fu">align_sequences</span>(vertebrates_busco_seqs, <span class="at">threads =</span> <span class="dv">4</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim alignment to remove columns with &gt;50% of gaps</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>aln_vertebrates_trimmed <span class="ot">&lt;-</span> <span class="fu">lapply</span>(aln_vertebrates, trim_alignment, <span class="at">max_gap =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s infer a species tree. As outgroup, we will use the yeast <em>Saccharomyces cerevisiae</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"saccharomyces.cerevisiae"</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get a single tree from a concatenated alignment</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>vertebrates_busco_trees <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">conc =</span> <span class="fu">concatenate_alignments</span>(aln_vertebrates_trimmed)),</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    outgroup, <span class="at">threads =</span> <span class="dv">6</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Save tree</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    vertebrates_busco_trees, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"trees"</span>, <span class="st">"vertebrates_busco_trees.rda"</span>)</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="obtaining-busco-scores" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="obtaining-busco-scores"><span class="header-section-number">1.4</span> Obtaining BUSCO scores</h2>
<p>Finally, since we ran BUSCO to obtain single-copy gene families, we will also use BUSCO’s output to explore gene space completeness across species in Ensembl instances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read BUSCO completeness stats</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Ensembl Fungi</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>fungi_busco_scores <span class="ot">&lt;-</span> <span class="fu">read_busco</span>(</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/busco_fungi/ensembl_fungi"</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Ensembl Plants</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>plants_busco_scores <span class="ot">&lt;-</span> <span class="fu">read_busco</span>(</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/busco_plants/ensemblplants"</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Ensembl Protists</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>protists_busco_scores <span class="ot">&lt;-</span> <span class="fu">read_busco</span>(</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/busco_protists/ensemblprotists"</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="do">## Ensembl Metazoa</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>metazoa_busco_scores <span class="ot">&lt;-</span> <span class="fu">read_busco</span>(</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/busco_metazoa/ensemblmetazoa"</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a><span class="do">## Ensembl Vertebrates</span></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>vertebrates_busco_scores <span class="ot">&lt;-</span> <span class="fu">read_busco</span>(</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/busco_vertebrates/ensemblvertebrates"</span></span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Save files</span></span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>    fungi_busco_scores, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>        <span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_scores"</span>, <span class="st">"fungi_busco_scores.rda"</span></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>    plants_busco_scores, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(</span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>        <span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_scores"</span>, <span class="st">"plants_busco_scores.rda"</span></span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>    protists_busco_scores, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(</span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>        <span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_scores"</span>, <span class="st">"protists_busco_scores.rda"</span></span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-49"><a href="#cb28-49" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb28-50"><a href="#cb28-50" aria-hidden="true" tabindex="-1"></a>    metazoa_busco_scores, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb28-51"><a href="#cb28-51" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(</span>
<span id="cb28-52"><a href="#cb28-52" aria-hidden="true" tabindex="-1"></a>        <span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_scores"</span>, <span class="st">"metazoa_busco_scores.rda"</span></span>
<span id="cb28-53"><a href="#cb28-53" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb28-54"><a href="#cb28-54" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-55"><a href="#cb28-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-56"><a href="#cb28-56" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb28-57"><a href="#cb28-57" aria-hidden="true" tabindex="-1"></a>    vertebrates_busco_scores, <span class="at">compress =</span> <span class="st">"xz"</span>,</span>
<span id="cb28-58"><a href="#cb28-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(</span>
<span id="cb28-59"><a href="#cb28-59" aria-hidden="true" tabindex="-1"></a>        <span class="st">"products"</span>, <span class="st">"result_files"</span>, <span class="st">"busco_scores"</span>, <span class="st">"vertebrates_busco_scores.rda"</span></span>
<span id="cb28-60"><a href="#cb28-60" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb28-61"><a href="#cb28-61" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="session-info" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="session-info">Session info</h2>
<p>This document was created under the following conditions:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       Ubuntu 22.04.3 LTS
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Brussels
 date     2024-02-12
 pandoc   3.1.1 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package          * version   date (UTC) lib source
 AnnotationDbi      1.64.1    2023-11-03 [1] Bioconductor
 ape                5.7-1     2023-03-13 [1] CRAN (R 4.3.2)
 aplot              0.2.2     2023-10-06 [1] CRAN (R 4.3.2)
 beeswarm           0.4.0     2021-06-01 [1] CRAN (R 4.3.2)
 Biobase            2.62.0    2023-10-24 [1] Bioconductor
 BiocFileCache      2.10.1    2023-10-26 [1] Bioconductor
 BiocGenerics     * 0.48.1    2023-11-01 [1] Bioconductor
 biomaRt          * 2.58.0    2023-10-24 [1] Bioconductor
 Biostrings       * 2.70.1    2023-10-25 [1] Bioconductor
 bit                4.0.5     2022-11-15 [1] CRAN (R 4.3.2)
 bit64              4.0.5     2020-08-30 [1] CRAN (R 4.3.2)
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.2)
 blob               1.2.4     2023-03-17 [1] CRAN (R 4.3.2)
 bold               1.3.0     2023-05-02 [1] CRAN (R 4.3.2)
 cachem             1.0.8     2023-05-01 [1] CRAN (R 4.3.2)
 cli                3.6.2     2023-12-11 [1] CRAN (R 4.3.2)
 codetools          0.2-19    2023-02-01 [4] CRAN (R 4.2.2)
 cogeqc           * 1.6.0     2023-10-24 [1] Bioconductor
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.2)
 conditionz         0.1.0     2019-04-24 [1] CRAN (R 4.3.2)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.2)
 crul               1.4.0     2023-05-17 [1] CRAN (R 4.3.2)
 curl               5.2.0     2023-12-08 [1] CRAN (R 4.3.2)
 data.table         1.14.10   2023-12-08 [1] CRAN (R 4.3.2)
 DBI                1.1.3     2022-06-18 [1] CRAN (R 4.3.2)
 dbplyr             2.4.0     2023-10-26 [1] CRAN (R 4.3.2)
 digest             0.6.33    2023-07-07 [1] CRAN (R 4.3.2)
 dplyr            * 1.1.4     2023-11-17 [1] CRAN (R 4.3.2)
 evaluate           0.23      2023-11-01 [1] CRAN (R 4.3.2)
 fansi              1.0.6     2023-12-08 [1] CRAN (R 4.3.2)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.2)
 filelock           1.0.3     2023-12-11 [1] CRAN (R 4.3.2)
 forcats          * 1.0.0     2023-01-29 [1] CRAN (R 4.3.2)
 foreach            1.5.2     2022-02-02 [1] CRAN (R 4.3.2)
 fs                 1.6.3     2023-07-20 [1] CRAN (R 4.3.2)
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.2)
 GenomeInfoDb     * 1.38.2    2023-12-13 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData   1.2.11    2023-12-21 [1] Bioconductor
 ggbeeswarm         0.7.2     2023-04-29 [1] CRAN (R 4.3.2)
 ggfun              0.1.3     2023-09-15 [1] CRAN (R 4.3.2)
 ggplot2          * 3.4.4     2023-10-12 [1] CRAN (R 4.3.2)
 ggplotify          0.1.2     2023-08-09 [1] CRAN (R 4.3.2)
 ggtree             3.10.0    2023-10-24 [1] Bioconductor
 glue               1.6.2     2022-02-24 [1] CRAN (R 4.3.2)
 gridGraphics       0.5-1     2020-12-13 [1] CRAN (R 4.3.2)
 gtable             0.3.4     2023-08-21 [1] CRAN (R 4.3.2)
 here             * 1.0.1     2020-12-13 [1] CRAN (R 4.3.2)
 Herper           * 1.10.1    2024-02-08 [1] Github (RockefellerUniversity/Herper@ae37f3d)
 hms                1.1.3     2023-03-21 [1] CRAN (R 4.3.2)
 htmltools          0.5.7     2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets        1.6.4     2023-12-06 [1] CRAN (R 4.3.2)
 httpcode           0.3.0     2020-04-10 [1] CRAN (R 4.3.2)
 httr               1.4.7     2023-08-15 [1] CRAN (R 4.3.2)
 igraph             2.0.1.1   2024-01-30 [1] CRAN (R 4.3.2)
 IRanges          * 2.36.0    2023-10-24 [1] Bioconductor
 iterators          1.0.14    2022-02-05 [1] CRAN (R 4.3.2)
 jsonlite           1.8.8     2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST           1.42.0    2023-10-24 [1] Bioconductor
 knitr              1.45      2023-10-30 [1] CRAN (R 4.3.2)
 lattice            0.22-5    2023-10-24 [4] CRAN (R 4.3.1)
 lazyeval           0.2.2     2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle          1.0.4     2023-11-07 [1] CRAN (R 4.3.2)
 lubridate        * 1.9.3     2023-09-27 [1] CRAN (R 4.3.2)
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.2)
 Matrix             1.6-3     2023-11-14 [4] CRAN (R 4.3.2)
 memoise            2.0.1     2021-11-26 [1] CRAN (R 4.3.2)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.2)
 nlme               3.1-163   2023-08-09 [4] CRAN (R 4.3.1)
 patchwork          1.2.0     2024-01-08 [1] CRAN (R 4.3.2)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.2)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.2)
 plyr               1.8.9     2023-10-02 [1] CRAN (R 4.3.2)
 png                0.1-8     2022-11-29 [1] CRAN (R 4.3.2)
 prettyunits        1.2.0     2023-09-24 [1] CRAN (R 4.3.2)
 progress           1.2.3     2023-12-06 [1] CRAN (R 4.3.2)
 purrr            * 1.0.2     2023-08-10 [1] CRAN (R 4.3.2)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.2)
 rappdirs           0.3.3     2021-01-31 [1] CRAN (R 4.3.2)
 Rcpp               1.0.11    2023-07-06 [1] CRAN (R 4.3.2)
 RCurl              1.98-1.13 2023-11-02 [1] CRAN (R 4.3.2)
 readr            * 2.1.4     2023-02-10 [1] CRAN (R 4.3.2)
 reshape2           1.4.4     2020-04-09 [1] CRAN (R 4.3.2)
 reticulate       * 1.35.0    2024-01-31 [1] CRAN (R 4.3.2)
 rjson              0.2.21    2022-01-09 [1] CRAN (R 4.3.2)
 rlang              1.1.2     2023-11-04 [1] CRAN (R 4.3.2)
 rmarkdown          2.25      2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot          2.0.4     2023-11-05 [1] CRAN (R 4.3.2)
 RSQLite            2.3.4     2023-12-08 [1] CRAN (R 4.3.2)
 rstudioapi         0.15.0    2023-07-07 [1] CRAN (R 4.3.2)
 S4Vectors        * 0.40.2    2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0     2023-11-28 [1] CRAN (R 4.3.2)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.2)
 stringi            1.8.3     2023-12-11 [1] CRAN (R 4.3.2)
 stringr          * 1.5.1     2023-11-14 [1] CRAN (R 4.3.2)
 taxize           * 0.9.100   2022-04-22 [1] CRAN (R 4.3.2)
 tibble           * 3.2.1     2023-03-20 [1] CRAN (R 4.3.2)
 tidyr            * 1.3.0     2023-01-24 [1] CRAN (R 4.3.2)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.2)
 tidytree           0.4.6     2023-12-12 [1] CRAN (R 4.3.2)
 tidyverse        * 2.0.0     2023-02-22 [1] CRAN (R 4.3.2)
 timechange         0.2.0     2023-01-11 [1] CRAN (R 4.3.2)
 treeio             1.26.0    2023-10-24 [1] Bioconductor
 tzdb               0.4.0     2023-05-12 [1] CRAN (R 4.3.2)
 utf8               1.2.4     2023-10-22 [1] CRAN (R 4.3.2)
 uuid               1.1-1     2023-08-17 [1] CRAN (R 4.3.2)
 vctrs              0.6.5     2023-12-01 [1] CRAN (R 4.3.2)
 vipor              0.4.7     2023-12-18 [1] CRAN (R 4.3.2)
 withr              2.5.2     2023-10-30 [1] CRAN (R 4.3.2)
 xfun               0.41      2023-11-01 [1] CRAN (R 4.3.2)
 XML                3.99-0.16 2023-11-29 [1] CRAN (R 4.3.2)
 xml2               1.3.6     2023-12-04 [1] CRAN (R 4.3.2)
 XVector          * 0.42.0    2023-10-24 [1] Bioconductor
 yaml               2.3.8     2023-12-11 [1] CRAN (R 4.3.2)
 yulab.utils        0.1.2     2023-12-22 [1] CRAN (R 4.3.2)
 zlibbioc           1.48.0    2023-10-24 [1] Bioconductor
 zoo                1.8-12    2023-04-13 [1] CRAN (R 4.3.2)

 [1] /home/faalm/R/x86_64-pc-linux-gnu-library/4.3
 [2] /usr/local/lib/R/site-library
 [3] /usr/lib/R/site-library
 [4] /usr/lib/R/library

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link" aria-label="Preface">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/chapter_02.html" class="pagination-link" aria-label="Identification and classification of duplicated genes in Ensembl and Ensembl Genomes">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Identification and classification of duplicated genes in Ensembl and Ensembl Genomes</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>